### PR TITLE
First Draft - Learning with Disagreement

### PR DESCRIPTION
**Pull request by vkakerbeck**
First version of a pytorch reimplementation of learning by using disagreement (original implementation here: https://github.com/pathak22/exploration-by-disagreement). This version learns comparatively well but there are still a few discrepancies in some of the training variables.
Use learn.py in the projects folder to test (example usage is in the script).

### CODE REVIEW

File: nupic/embodied/agents/curious_ppo_agent.py [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+    envs : [VecEnv]
+        List of vector enviornments to use for experience collection.
+
+    """
+
+    envs = None
+
+    def __init__(
+        self,
+        *,
```
Comment (lucasosouza) [inline nupic/embodied/agents/curious_ppo_agent.py:80]: What does the asterisk stands for here?

File: nupic/embodied/agents/curious_ppo_agent.py [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+            List of environments, optionally with wrappers etc.
+        dynamics_list : [Dynamics]
+            List of dynamics models.
+        nlump : int
+            ..
+
+        """
+        # Specify parameters that should be optimized
+        param_list = (
+            self.stochpol.param_list + self.dynamics_list[0].auxiliary_task.param_list
```
Comment (lucasosouza) [inline nupic/embodied/agents/curious_ppo_agent.py:143]: That is interesting. Why only the param_list of auxiliary task for the first task of the dynamic list, and not the others?

File: nupic/embodied/agents/curious_ppo_agent.py [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+            param_list = param_list + dynamic.param_list
+
+        # Initialize the optimizer
+        self.optimizer = torch.optim.Adam(param_list, lr=self.lr)
+        # Set gradients to zero
+        self.optimizer.zero_grad()
+
+        # set parameters
+        self.nenvs = nenvs = len(env_fns)
+        self.nlump = nlump
```
Comment (lucasosouza) [inline nupic/embodied/agents/curious_ppo_agent.py:155]: I can't figure out what lump means, it is not is the docstring either.

File: nupic/embodied/agents/curious_ppo_agent.py [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+            int_rew_coeff=self.int_coeff,
+            ext_rew_coeff=self.ext_coeff,
+            dynamics_list=dynamics_list,
+        )
+
+        # Initialize buffers for advantages and returns of each rollout
+        self.buf_advs = np.zeros((nenvs, self.rollout.nsteps), np.float32)
+        self.buf_rets = np.zeros((nenvs, self.rollout.nsteps), np.float32)
+
+        if self.normrew:
```
Comment (lucasosouza) [inline nupic/embodied/agents/curious_ppo_agent.py:185]: There are too many variable abbreviations used, which makes the code hard to read for someone unfamiliar with it. 

Abbreviations are generally ok if the meaning is obvious, but not all of them are. For example we could change `normrew` to `normalize_rewards`, or ent_coef to `entropy_coef`or ext_coeff to `external_reward_coef`. The code should preferably read like a paper. That is going to help not only reading the code, and associating it with the method described in the paper, but also maintaining the code (when we need to search for a variable for example).

File: nupic/embodied/agents/curious_ppo_agent.py [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+        use_news : bool
+            Whether to use news (which are the done infos from the environment).
+        gamma : float
+            Discount factor for the rewards.
+        lam : float
+            Generalized advantage estimator smoothing parameter.
+
+        """
+        nsteps = self.rollout.nsteps
+        lastgaelam = 0
```
Comment (lucasosouza) [inline nupic/embodied/agents/curious_ppo_agent.py:216]: Also couldn't figure out what lastgaelam means.

File: nupic/embodied/agents/curious_ppo_agent.py [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+        for t in range(nsteps - 1, -1, -1):  # going backwards from last step in seg
+            # Is the next step a new run (done=True)?
+            nextnew = (
+                self.rollout.buf_news[:, t + 1]
+                if t + 1 < nsteps
+                else self.rollout.buf_new_last
+            )
+            if not use_news:
+                nextnew = 0
+            nextnotnew = 1 - nextnew
```
Comment (lucasosouza) [inline nupic/embodied/agents/curious_ppo_agent.py:226]: What does nextnotnew stands for?

File: nupic/embodied/agents/curious_ppo_agent.py [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+                # Get the entropy of the current policy
+                entropy = torch.mean(self.stochpol.pd.entropy())
+                # Get the value estimate of the policies value head
+                vpred = self.stochpol.vpred
+                # Calculate the msq difference between value estimate and return
+                vf_loss = 0.5 * torch.mean(
+                    (vpred.squeeze() - torch.tensor(rets).to(self.device)) ** 2
+                )
+                # Put old nlps from buffer into tensor
+                nlps = torch.tensor(flatten_dims(nlps, 0)).to(self.device)
```
Comment (lucasosouza) [inline nupic/embodied/agents/curious_ppo_agent.py:366]: What are nlps?

File: nupic/embodied/envs/real_robots/policies.py [hunk 8]
```diff
@@ -20,8 +20,11 @@
 #
 # ------------------------------------------------------------------------------
 import numpy as np
-from real_robots.policy import BasePolicy
```
Comment (lucasosouza) [inline nupic/embodied/envs/real_robots/policies.py:23]: Wouldn't that break import the class altogether, since it inherits from BasePolicy?

File: nupic/embodied/envs/rollout.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+        self.ac_space = ac_space
+        self.nlumps = nlumps
+        self.lump_stride = nenvs // self.nlumps
+        self.envs = envs
+        self.policy = policy
+        self.dynamics_list = dynamics_list
+
+        # Define the reward function as a weighted combination of internal and (clipped)
+        # external rewards.
+        self.reward_fun = (
```
Comment (lucasosouza) [inline nupic/embodied/envs/rollout.py:99]: I am not a big fan of using lambdas for currying. The reason is if you redefine the variables in the global scope (in this case ext_rew_coeef and int_rew_coeff), they will change the output of lambda. If that is not the intended behavior, it would be safer to use partial instead. 

For an example see the second answer in this thread: https://stackoverflow.com/questions/3252228/python-why-is-functools-partial-necessary

File: nupic/embodied/envs/vec_env.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
+"""
```
Comment (lucasosouza) [inline nupic/embodied/envs/vec_env.py:1]: It is missing the license header (should be in all files, please extend this comment to all files)

File: nupic/embodied/envs/wrappers.py [hunk 11]
```diff
@@ -0,0 +1,539 @@
...
+            low=0, high=255, shape=(84, 84, 1), dtype=np.uint8
+        )
+
+    def observation(self, obs):
+        return ProcessFrame84.process(obs, crop=self.crop)
+
+    @staticmethod
+    def process(frame, crop=True):
+        if frame.size == 210 * 160 * 3:
+            img = np.reshape(frame, [210, 160, 3]).astype(np.float32)
```
Comment (lucasosouza) [inline nupic/embodied/envs/wrappers.py:68]: We can refactor this to be a single if, that will make it more flexible to be extended. For example:
```
valid_frame_sizes = [(210,160,3), (250, 160, 3), (224, 240, 3)]
fs_mapping = {np.prod(fs):fs for fs in valid_frame_sizes}
if frame.size in fs_mapping:
    img = np.reshape(frame, fs_mapping[frame.size]).astype(np.float32)
```

With that, we can go further and make valid_frame_sizes a constant at a class level. So process would have to become a classmethod to access it

```
class ProcessFrame84():
   VALID_FRAME_SIZES = [...]
   
   def __init__(self):

...

@classmethod
def process(cls, frame, crop=True):
    cls.valid_frames_sizes
```

File: nupic/embodied/envs/wrappers.py [hunk 12]
```diff
@@ -0,0 +1,539 @@
...
+
+    def reset(self):
+        """Clear past frame buffer and init. to first obs. from inner env."""
+        self._obs_buffer.clear()
+        obs = self.env.reset()
+        self._obs_buffer.append(obs)
+        return obs
+
+
+class ProcessFrame84(gym.ObservationWrapper):
```
Comment (lucasosouza) [inline nupic/embodied/envs/wrappers.py:54]: What is this class being used for? There are a lot of constants scattered across the methods, which makes this very specific to something. A one line docstringer would be a good addition to clarify that.

File: nupic/embodied/envs/wrappers.py [hunk 13]
```diff
@@ -0,0 +1,539 @@
...
+        elif frame.size == 224 * 240 * 3:  # mario resolution
+            img = np.reshape(frame, [224, 240, 3]).astype(np.float32)
+        else:
+            assert False, "Unknown resolution." + str(frame.size)
+        img = img[:, :, 0] * 0.299 + img[:, :, 1] * 0.587 + img[:, :, 2] * 0.114
+        size = (84, 110 if crop else 84)
+        resized_screen = np.array(
+            Image.fromarray(img).resize(size, resample=Image.BILINEAR), dtype=np.uint8
+        )
+        x_t = resized_screen[18:102, :] if crop else resized_screen
```
Comment (lucasosouza) [inline nupic/embodied/envs/wrappers.py:80]: It would be good to have some more background on what do this constants represent. We could either assign them to variables with a meaningful descriptive name or add a comment instead if assigning them to a name doesn't make sense.

File: nupic/embodied/envs/wrappers.py [hunk 14]
```diff
@@ -0,0 +1,539 @@
...
+class NoReward(gym.Wrapper):
+    def __init__(self, env):
+        gym.Wrapper.__init__(self, env)
+
+    def step(self, action):
+        ob, rew, done, info = self.env.step(action)
+        return ob, 0.0, done, info
+
+
+def make_multi_pong(frame_stack=True):
```
Comment (lucasosouza) [inline nupic/embodied/envs/wrappers.py:329]: We should move all functions to after the classes (or before). I believe the autoformatter will do that for you.

File: nupic/embodied/envs/wrappers.py [hunk 15]
```diff
@@ -0,0 +1,539 @@
...
+                self._env.robot.jdict["finger10_to_finger11_joint"].get_position(),
+            ]
+        )
+        # Currently orientation always makes gripper point down -> could be
+        # changed later and added into action space. (TODO)
+        desiredOrientation = pybullet.getQuaternionFromEuler([0, 3.14, -1.57])
+        cartesian[:3] = loc
+        cartesian[3:] = desiredOrientation
+
+        if action == 0:  # forward
```
Comment (lucasosouza) [inline nupic/embodied/envs/wrappers.py:471]: For readability, what about mapping action to (cartesian_pos, increment_sign) separately from the assignment operation then doing
```
cartesian_pos, increment_sign = action_to_movement[action]
if action in range(5):
    cartesian[cartesian_pos] += self.increment_size * sign
else:
    gripper = grip + self.increment_size * sign
```

File: nupic/embodied/policies/auxiliary_tasks.py [hunk 16]
```diff
@@ -0,0 +1,479 @@
...
+        obs = (obs - self.ob_mean) / self.ob_std
+        # Reshape observations
+        obs = np.transpose(obs, [i for i in range(len(obs.shape) - 3)] + [-1, -3, -2])
+        # Get features from the features_model
+        act = self.features_model(torch.tensor(obs).to(self.device))
+
+        if has_timesteps:
+            act = unflatten_first_dim(act, sh)
+
+        """ print(
```
Comment (lucasosouza) [inline nupic/embodied/policies/auxiliary_tasks.py:163]: To make the print statements permanent, without having to comment, an option is to add a verbose flag and put the print statements under the flag.

File: nupic/embodied/policies/auxiliary_tasks.py [hunk 17]
```diff
@@ -0,0 +1,479 @@
...
+        the hidden activations, followed by another fc layer outputting the predicted
+        action."""
+        self.fc = torch.nn.Sequential(
+            torch.nn.Linear(self.feat_dim * 2, self.policy.hid_dim),
+            torch.nn.ReLU(),
+            torch.nn.Linear(self.policy.hid_dim, self.ac_space.n),
+        ).to(self.device)
+        # Add auxiliary task output layers to list of optimized parameters
+        self.param_list = self.param_list + [dict(params=self.fc.parameters())]
+        # Initialite weights with xavier_uniform and constant biases
```
Comment (lucasosouza) [inline nupic/embodied/policies/auxiliary_tasks.py:235]: typo, Initialize

File: nupic/embodied/policies/curious_cnn_policy.py [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+            torch.nn.Linear(self.hid_dim, self.hid_dim),
+            torch.nn.ReLU(),
+        ).to(self.device)
+        # policy and value function head of the policy network.
+        self.pd_head = torch.nn.Linear(self.hid_dim, pdparamsize).to(self.device)
+        self.vf_head = torch.nn.Linear(self.hid_dim, 1).to(self.device)
+
+        # Define parameters to be optimized
+        self.param_list = [
+            dict(params=self.features_model.parameters()),
```
Comment (lucasosouza) [inline nupic/embodied/policies/curious_cnn_policy.py:123]: This format is only required if you are specifying individual arguments (e.g. learning rate, momentum) for separate set of parameters. 

If not, you can just concatenate a list of Variables instead. So just make param_list a list, then you can call extend: `self.param_list.extend(new_param_list)`. Or during initialization use the spread operator:
```
self.param_list = [*self.features_model.parameters(), *self.pd_hidden.parameters(), ...]
```

File: nupic/embodied/utils/model_parts.py [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+        # Add fc layer at end for feature output
+        self.fc = torch.nn.Linear(self.flatten_dim, feat_dim).to(device)
+
+        self.last_nonlinear = last_nonlinear
+        self.layernormalize = layernormalize
+        # initialize weight with xavier uniform distribution
+        self.init_weight()
+
+    def init_weight(self):
+        """Initialize wieght of network."""
```
Comment (lucasosouza) [inline nupic/embodied/utils/model_parts.py:135]: typo, weight

File: nupic/embodied/utils/model_parts.py [hunk 20]
```diff
@@ -0,0 +1,243 @@
+import torch
+import numpy as np
+
+
+def flatten_dims(x, dim):
+    """Flatten dimension dim of x.
```
Comment (lucasosouza) [inline nupic/embodied/utils/model_parts.py:6]: I love numpy style docstrings, they are very clean. But in nupic we use the more traditional sphinx style docstring. At some point would be good to standardize, but it doesn't really matter now if we have mixed types of docstrings for a while.

File: nupic/embodied/utils/model_parts.py [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+        Returns
+        -------
+        array
+            Output features of the network.
+
+        """
+        # run x through the convolutional layers
+        x = self.conv(x)
+        # Get flattened version of conv output
+        # TODO: check that contiguous.view actually does what its supposed to
```
Comment (lucasosouza) [inline nupic/embodied/utils/model_parts.py:160]: torch now has a nn.Flatten() layer that can be added to the sequential

File: nupic/embodied/utils/model_parts.py [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+        x = self.fc(x)
+        if self.last_nonlinear is not None:
+            # apply activation function the the last layer
+            x = self.last_nonlinear(x)
+        if self.layernormalize:
+            # normalize the last layer
+            x = self.layernorm(x)
+        return x
+
+    def layernorm(self, x):
```
Comment (lucasosouza) [inline nupic/embodied/utils/model_parts.py:174]: torch also has layernorm, see if nn.LayerNorm does the trick

File: nupic/embodied/utils/model_parts.py [hunk 23]
```diff
@@ -0,0 +1,243 @@
...
+        )  # dims is calculated manually, 84*84 -> 20*20 -> 9*9 ->7*7
+        # run through fully connected layer
+        x = self.fc(x)
+        if self.last_nonlinear is not None:
+            # apply activation function the the last layer
+            x = self.last_nonlinear(x)
+        if self.layernormalize:
+            # normalize the last layer
+            x = self.layernorm(x)
+        return x
```
Comment (lucasosouza) [inline nupic/embodied/utils/model_parts.py:172]: If we change all operations to be layer, we can make the classifier be a sequential as well, so `self.classifier = nn.Sequential(nn.Flatten, fc, layernorm)` and then just call self.classifier(x)

File: nupic/embodied/utils/model_parts.py [hunk 24]
```diff
@@ -0,0 +1,243 @@
...
+        for m in self.deconv:
+            if isinstance(m, torch.nn.ConvTranspose2d):
+                torch.nn.init.xavier_uniform_(m.weight.data)
+                torch.nn.init.constant_(m.bias.data, 0.0)
+
+    def forward(self, features):
+        x = self.fc(features)
+        x = x.view((-1,) + self.sh)
+        x = self.deconv(x)
+        x = x[:, :, 6:-6, 6:-6]
```
Comment (lucasosouza) [inline nupic/embodied/utils/model_parts.py:239]: 6:-6 seems very specific, what is it about? Can it be a function of some other number defined elsewhere in the class? I am always wary of hard-coded constants as they can bring some hidden bugs (as in some other constant is changed and then this becomes obsolete).

**Review by lucasosouza — APPROVED** [review]
I'm looking forward to run it! It is a big PR, so I left a lot of comments. They are all optional and meant to be helpful. At this point to move forward we should just merge it as it is, since it doesn't any other existing code, and then implement changes as an additional PR.

In general, this all looks great. I did have a hard time reading it, specially the curious_ppo_agent file, due to the abbreviations. We should strive to make the variable names meaningful so the code can read like a paper and easily understood by someone not working in the project directly. 

I don't know how much of this code was copied as it is from another source. Any code copied should clearly be identified with a url to the original implementation in the module docstring or the class/method docstring. 

One other change worthy of note: all files need the license disclaimer at the beginning.
